streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -583,16 +583,21 @@ private void runLoop() {
                 }
             }
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
+            // try to process several records from each task via the topology, and also trigger punctuate
             // functions if necessary, which may result in more records going through the topology in this loop
             if (totalNumBuffered > 0 || polledRecords) {
                 totalNumBuffered = 0;
 
                 if (!activeTasks.isEmpty()) {
                     for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
-
+                        // process this task's records to completion before
+                        // context switching to another task
+                        while (true) {
```

Comment:
perhaps this could be:
```
int numProcessed;
while ((numProcessed = task.process()) != 0) {
..}
```
I just have an aversion to `while(true) ... break`